ddht/v5_1/client.py

```diff
@@ -400,9 +400,17 @@ async def find_nodes(
                 return responses
 
     async def talk(
-        self, endpoint: Endpoint, node_id: NodeID, protocol: bytes, request: bytes
+        self, endpoint: Endpoint, node_id: NodeID, protocol: bytes, payload: bytes
     ) -> InboundMessage[TalkResponseMessage]:
-        raise NotImplementedError
+        with self._get_request_id(node_id) as request_id:
+            request = AnyOutboundMessage(
+                TalkRequestMessage(request_id, protocol, payload), endpoint, node_id,
+            )
+            async with self.dispatcher.subscribe_request(
+                request, TalkResponseMessage
+            ) as subscription:
+                with trio.fail_after(REQUEST_RESPONSE_TIMEOUT):
```

Comment:
trio and context managers are both such beautiful additions to the language

Reply:
There's some duplication of work here. `subscribe_request` calls `_manage_request_response` which has it's own call to `trio.move_on_after(REQUEST_RESPONSE_TIMEOUT)`. This will probably be confusing to the next person who tries to increase one of these timeouts and finds that the timeout has not increased.

I'm not entirely sure where it belongs:
- part of me thinks the timeout belongs in `_manage_request_response`, and that way there's a uniform timeout all requests use and callers don't need to worry about timeouts
- the other part of me thinks it belongs up here, and that the timeout in `_manage_request_response` ought to be removed. This gives callers the flexibility to change the timeout and I have the feeling they'll sometimes want to. It also makes things more explicit for someone browsing the code and who wonders what `talk` does and how it handles requests which come in too late.

Reply:
I'll look into seeing if the timeout in `_manage_request_response`, which will result in the channel being closed, will cause the call on the next line to `await subscription.receive()` to throw something like `trio.BrokenResourceError` which can then either be allowed to bubble, or potentially we can raise an internal error of some kind.

Reply:
Alright.  I was able to remove all of these second-layer timeouts and rely solely on the internal timeout that arises from `_manage_request_response`.  Now these methods raise `trio.EndOfChannel` which I think we'll want to eventually replace with a custom internal exception.  I originally had it re-raising `trio.TooSlowError` but I reverted that since it's possible that the channel gets closed due to some other mechanism, so eventually we may want to re-raise an internal exception that is more generic like `RequestFailure`